### PR TITLE
Use Cache for bonding peers

### PR DESCRIPTION
**Pull request by davemec**
## PR description
Updated to use a Cache for the bonding peers collection.

## Fixed Issue(s)
#792

### CODE REVIEW

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java [hunk 1]
```diff
@@ -586,6 +590,17 @@ public void setRetryDelayFunction(final RetryDelayFunction retryDelayFunction) {
     this.retryDelayFunction = retryDelayFunction;
   }
 
+  public void handleBondingRequest(final DiscoveryPeer peer) {
+    final DiscoveryPeer peerToBond =
+        peerTable.get(peer).filter(known -> known.discoveryEndpointMatches(peer)).orElse(peer);
```
Comment (mbaxter) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:595]: Shouldn't we also check against the `bondingPeers` collection?  

Maybe we need a helper method `resolvePeer` that takes a peer and replaces it with representations from the `peerTable` or `bondingPeers` collection as needed.  Then we could dedupe the same logic in `onMessage`.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java [hunk 2]
```diff
@@ -586,6 +590,17 @@ public void setRetryDelayFunction(final RetryDelayFunction retryDelayFunction) {
...
+
+    if (peerPermissions.allowOutboundBonding(peerToBond)
+        && !PeerDiscoveryStatus.BONDED.equals(peerToBond.getStatus())
+        && !PeerDiscoveryStatus.BONDING.equals(peerToBond.getStatus())) {
```
Comment (mbaxter) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:599]: (nit) I think we could simplify:
```suggestion
    if (peerPermissions.allowOutboundBonding(peerToBond)
        && .peerToBond.getStatus().equals(PeerDiscoveryStatus.KNOWN)) {
```

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java [hunk 3]
```diff
@@ -107,7 +108,8 @@
   private static final int PEER_REFRESH_ROUND_TIMEOUT_IN_SECONDS = 5;
   protected final TimerUtil timerUtil;
   private final PeerTable peerTable;
-  private final Map<Bytes, DiscoveryPeer> bondingPeers;
+  private final Cache<Bytes, DiscoveryPeer> bondingPeers =
+      CacheBuilder.newBuilder().maximumSize(50).expireAfterWrite(10, TimeUnit.MINUTES).build();
```
Comment (mbaxter) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:112]: good catch!

Comment (davemec) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:112]: Thanks to @shemnon

**Pull request merged**